File: cpp/KalkCpp/Shape/Shape3D/Polyhedron/cube.cpp
#include "cube.h"
#include "math.h"
#include "algorithm"
#include "../../Shape2D/Polygon/Quadrilateral/square.h"



Cube::Cube(const std::vector<Point3D>& vp3d, Square* bs, ColorRGB* color): Polyhedron(vp3d, bs, "Cube", color)
{

}


Cube* Cube::clone() const
{
    return new Cube(*this);
}

std::vector<Point3D> Cube::buildCube(Square* s) const
{
    std::vector<Point3D> points = std::vector<Point3D>();

    for(std::vector<Point2D>::const_iterator it = s->getPoints().begin(); it != s->getPoints().end(); ++it)
    {
        points.push_back(Point3D::p2dTop3d(*it));
    }

    for(std::vector<Point2D>::const_iterator it = s->getPoints().begin(); it != s->getPoints().end(); ++it)
    {
        Point2D p = *it;
        points.push_back(Point3D(p.getX(), p.getY(), s->base()));
    }

    return points;
}



Cube::Cube(Square* bs, ColorRGB* color): Polyhedron(buildCube(bs), bs, "Cube", color)
{

}



Cube* Cube::operator-(const Shape3D& s) const
{
    Square* newbaseShape = static_cast<Square*>(baseShape)->operator -(*s.getBaseShape());
    Cube* newCube = new Cube(buildCube(newbaseShape),newbaseShape, getColor()->operator -(*s.getColor()));
    return newCube;
}



Cube* Cube::operator+(const Shape3D& s) const
{
    Square* newbaseShape = static_cast<Square*>(baseShape)->operator +(*s.getBaseShape());
    Cube* newCube = new Cube(buildCube(newbaseShape),newbaseShape, getColor()->operator +(*s.getColor()));
    return newCube;
}



Cube* Cube::operator*(const Shape3D& s) const
{
    Square* newbaseShape = static_cast<Square*>(baseShape)->operator *(*s.getBaseShape());
    Cube* newCube = new Cube(buildCube(newbaseShape),newbaseShape, getColor()->operator *(*s.getColor()));
    return newCube;
}



Cube* Cube::operator/(const Shape3D& s) const
{
    Square* newbaseShape = static_cast<Square*>(baseShape)->operator /(*s.getBaseShape());
    Cube* newCube = new Cube(buildCube(newbaseShape),newbaseShape, getColor()->operator /(*s.getColor()));
    return newCube;
}



unsigned int Cube::getNumberCorner() const
{
    return 12;
}



unsigned int Cube::getNumberFaces() const
{
    return getNumberCorner() / 2;
}



double Cube::volume() const
{
    Square* tmp = static_cast<Square*>(baseShape);
    return pow(tmp->base(),3);
}



double Cube::lateralSurface() const
{
  Square* tmp = static_cast<Square*>(baseShape);
  return 4 * pow(tmp->base(), 2);
}



double Cube::totalSurface() const
{
  Square* tmp = static_cast<Square*>(baseShape);
  return 6 * pow(tmp->base(), 2);
}

double Cube::diagonal() const
{
    return side() * sqrt(3);
}

double Cube::side() const
{
    Square* tmp = static_cast<Square*>(baseShape);
    return tmp->base();
}


Cube::~Cube()
{

}

File: Java/src/Shape/Shape3D/Polyhedron/Cube.java
package Shape.Shape3D.Polyhedron;

import Color.ColorRGB;
import Shape.Point2D;
import Shape.Point3D;
import Shape.Shape2D.Polygon.Quadrilateral.Square;
import Shape.Shape3D.Shape3D;

import java.util.ArrayList;

public class Cube extends Polyhedron {

    public Cube(ArrayList<Point3D> p, Square bShape) {
        super(p, bShape, "Cube");
        points = buildCube(bShape);
    }

    public Cube(ArrayList<Point3D> p, Square bShape, ColorRGB color) {
        super(p, bShape, "Cube", color);
        points = buildCube(bShape);
    }

    public Cube(Square s) {
        super(null, s, "Cube");
        points = buildCube(s);
    }

    public Cube(Square s, ColorRGB color) {
        super(null, s, "Cube", color);
        points = buildCube(s);
    }

    public Cube(Cube c) {
        super(c);
    }

    private ArrayList<Point3D> buildCube(Square s) {
        ArrayList<Point3D> points = new ArrayList<>();

        for (Point2D point2d : s.getPoints()) {
            points.add(Point3D.p2dTop3d(point2d));
        }

        for (Point2D point2d : s.getPoints()) {
            points.add(new Point3D(point2d.getX(), point2d.getY(), s.base()));
        }
        return points;
    }

    public Cube sum(Shape3D other) {
        Square newbaseShape = (Square) baseShape.sum(other.getBaseShape());
        return new Cube(
                buildCube(newbaseShape), newbaseShape, getColor().sum(other.getColor()));
    }


    public Cube sub(Shape3D other) {
        Square newbaseShape = (Square) baseShape.sub(other.getBaseShape());
        return new Cube(
                buildCube(newbaseShape), newbaseShape, getColor().sub(other.getColor()));
    }

    public Cube mul(Shape3D other) {
        Square newbaseShape = (Square) baseShape.mul(other.getBaseShape());
        return new Cube(
                buildCube(newbaseShape), newbaseShape, getColor().mul(other.getColor()));
    }

    public Cube div(Shape3D other) {
        Square newbaseShape = (Square) baseShape.div(other.getBaseShape());
        return new Cube(
                buildCube(newbaseShape), newbaseShape, getColor().div(other.getColor()));
    }

    public int getNumberCorner() {
        return 12;
    }

    public int getNumberFaces() {
        return getNumberCorner() / 2;
    }


    public double volume() {
        Square s = (Square) baseShape;
        return Math.pow(s.base(), 3);
    }


    public double lateralSurface() {
        Square s = (Square) baseShape;
        return 4 * Math.pow(s.base(), 2);
    }

    public double totalSurface() {
        Square s = (Square) baseShape;
        return 6 * Math.pow(s.base(), 2);
    }

    public double diagonal() {
        return side() * Math.sqrt(3);
    }

    double side() {
        Square s = (Square) baseShape;
        return s.base();
    }

    public String toString() {
        return super.toString() + "Side " + side() + "\nVolume " + volume() +
                "\nLateralSurface " + lateralSurface() + "\nTotalSurface " + totalSurface();
    }

}
